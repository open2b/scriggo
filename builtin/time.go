// Copyright 2021 The Scriggo Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

// Copyright 2009 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package builtin

import (
	"fmt"
	"time"

	"github.com/open2b/scriggo/native"
)

// A Duration represents the elapsed time between two instants
// as an int64 nanosecond count. The representation limits the
// largest representable duration to approximately 290 years.
type Duration = time.Duration

// A Time represents an instant in time.
//
// It is a stripped down version of the Go time.Time type, with additional
// methods to show time values in JavaScript and JSON contexts.
type Time struct {
	t time.Time
}

// NewTime returns a Time value at the time of t. It is not intended to be
// used as a builtin but it can be used to implement builtins that return time
// values.
//
// For example, a builtin that returns the current local time rounded to
// milliseconds and with location "America/New_York" can be implemented as
//
//	func now() builtin.Time {
//	    l := time.LoadLocation("America/New_York")
//	    t := time.Now().Rounded(time.Millisecond).In(l)
//	    return builtin.NewTime(t)
//	}
//
// and once added to the declarations
//
//	native.Declarations{
//	    ...
//	    "now" : now,
//	    ...
//	}
//
// it can be used in a template
//
//	Current time: {{ now() }}
func NewTime(t time.Time) Time {
	return Time{t}
}

// Add returns the time t+d.
func (t Time) Add(d Duration) Time {
	return Time{t.t.Add(d)}
}

// AddDate returns the time corresponding to adding the given number of years,
// months and days to t. For example, AddDate(-1, 2, 3) applied to January 1,
// 2011 returns March 4, 2010.
//
// AddDate normalizes its result in the same way that the Date function does.
func (t Time) AddDate(years, months, days int) Time {
	return Time{t.t.AddDate(years, months, days)}
}

// After reports whether the time instant t is after u.
func (t Time) After(u Time) bool {
	return t.t.After(u.t)
}

// Before reports whether the time instant t is before u.
func (t Time) Before(u Time) bool {
	return t.t.Before(u.t)
}

// Clock returns the hour, minute and second within the day specified by t.
// hour is in the range [0, 23], minute and second are in the range [0, 59].
func (t Time) Clock() (hour, minute, second int) {
	return t.t.Clock()
}

// Date returns the year, month and day in which t occurs. month is in the
// range [1, 12] and day is in the range [1, 31].
func (t Time) Date() (year, month, day int) {
	y, m, d := t.t.Date()
	return y, int(m), d
}

// Day returns the day of the month specified by t, in the range [1, 31].
func (t Time) Day() int {
	return t.t.Day()
}

// Equal reports whether t and u represent the same time instant.
func (t Time) Equal(u Time) bool {
	return t.t.Equal(u.t)
}

// Format returns a textual representation of the time value formatted
// according to layout, which defines the format by showing how the reference
// time
//
//	Mon Jan 2 15:04:05 -0700 MST 2006
//
// would be displayed if it were the value; it serves as an example of the
// desired output. The same display rules will then be applied to the time
// value.
//
// A fractional second is represented by adding a period and zeros
// to the end of the seconds section of layout string, as in "15:04:05.000"
// to format a time stamp with millisecond precision.
func (t Time) Format(layout string) string {
	return t.t.Format(layout)
}

// Hour returns the hour within the day specified by t, in the range [0, 23].
func (t Time) Hour() int {
	return t.t.Hour()
}

// IsZero reports whether t represents the zero time instant,
// January 1, year 1, 00:00:00 UTC.
func (t Time) IsZero() bool {
	return t.t.IsZero()
}

// JS returns the time as a JavaScript date. The result is undefined if the
// year of t is not in the range [-999999, 999999].
func (t Time) JS() native.JS {
	y := t.t.Year()
	ms := int64(t.t.Nanosecond()) / int64(time.Millisecond)
	name, offset := t.t.Zone()
	if name == "UTC" {
		format := `new Date("%0.4d-%0.2d-%0.2dT%0.2d:%0.2d:%0.2d.%0.3dZ")`
		if y < 0 || y > 9999 {
			format = `new Date("%+0.6d-%0.2d-%0.2dT%0.2d:%0.2d:%0.2d.%0.3dZ")`
		}
		return native.JS(fmt.Sprintf(format, y, t.t.Month(), t.t.Day(), t.t.Hour(), t.t.Minute(), t.t.Second(), ms))
	}
	zone := offset / 60
	h, m := zone/60, zone%60
	if m < 0 {
		m = -m
	}
	format := `new Date("%0.4d-%0.2d-%0.2dT%0.2d:%0.2d:%0.2d.%0.3d%+0.2d:%0.2d")`
	if y < 0 || y > 9999 {
		format = `new Date("%+0.6d-%0.2d-%0.2dT%0.2d:%0.2d:%0.2d.%0.3d%+0.2d:%0.2d")`
	}
	return native.JS(fmt.Sprintf(format, y, t.t.Month(), t.t.Day(), t.t.Hour(), t.t.Minute(), t.t.Second(), ms, h, m))
}

// JSON returns a time in a format suitable for use in JSON.
func (t Time) JSON() native.JSON {
	return native.JSON(`"` + t.t.Format(time.RFC3339) + `"`)
}

// Month returns the month of the year specified by t, in the range [1, 12]
// where 1 is January and 12 is December.
func (t Time) Month() int {
	return int(t.t.Month())
}

// Minute returns the minute offset within the hour specified by t, in the
// range [0, 59].
func (t Time) Minute() int {
	return t.t.Minute()
}

// Nanosecond returns the nanosecond offset within the second specified by t,
// in the range [0, 999999999].
func (t Time) Nanosecond() int {
	return t.t.Nanosecond()
}

// Round returns the result of rounding t to the nearest multiple of d (since
// the zero time). The rounding behavior for halfway values is to round up.
// If d <= 0, Round returns t stripped of any monotonic clock reading but
// otherwise unchanged.
//
// Round operates on the time as an absolute duration since the zero time; it
// does not operate on the presentation form of the time. Thus, Round(Hour)
// may return a time with a non-zero minute, depending on the time's Location.
func (t Time) Round(d Duration) Time {
	return Time{t.t.Round(d)}
}

// Second returns the second offset within the minute specified by t, in the
// range [0, 59].
func (t Time) Second() int {
	return t.t.Second()
}

// String returns the time formatted.
func (t Time) String() string {
	return t.t.String()
}

// Sub returns the duration t-u. If the result exceeds the maximum (or minimum)
// value that can be stored in a Duration, the maximum (or minimum) duration
// will be returned.
// To compute t-d for a duration d, use t.Add(-d).
func (t Time) Sub(u Time) Duration {
	return t.t.Sub(u.t)
}

// Truncate returns the result of rounding t down to a multiple of d (since
// the zero time). If d <= 0, Truncate returns t stripped of any monotonic
// clock reading but otherwise unchanged.
//
// Truncate operates on the time as an absolute duration since the zero time;
// it does not operate on the presentation form of the time. Thus,
// Truncate(Hour) may return a time with a non-zero minute, depending on the
// time's Location.
func (t Time) Truncate(d Duration) Time {
	return Time{t.t.Truncate(d)}
}

// UTC returns t with the location set to UTC.
func (t Time) UTC() Time {
	return Time{t.t.UTC()}
}

// Unix returns t as a Unix time, the number of seconds elapsed since January
// 1, 1970 UTC. The result does not depend on the location associated with t.
// Unix-like operating systems often record time as a 32-bit count of seconds,
// but since the method here returns a 64-bit value it is valid for billions
// of years into the past or future.
func (t Time) Unix() int64 {
	return t.t.Unix()
}

// UnixNano returns t as a Unix time, the number of nanoseconds elapsed since
// January 1, 1970 UTC. The result is undefined if the Unix time in
// nanoseconds cannot be represented by an int64 (a date before the year 1678
// or after 2262). Note that this means the result of calling UnixNano on the
// zero Time is undefined. The result does not depend on the location
// associated with t.
func (t Time) UnixNano() int64 {
	return t.t.UnixNano()
}

// Weekday returns the day of the week specified by t.
func (t Time) Weekday() int {
	return int(t.t.Weekday())
}

// Year returns the year in which t occurs.
func (t Time) Year() int {
	return t.t.Year()
}

// YearDay returns the day of the year specified by t, in the range [1, 365]
// for non-leap years, and [1, 366] in leap years.
func (t Time) YearDay() int {
	return t.t.YearDay()
}
