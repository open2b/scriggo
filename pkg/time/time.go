// Go version: go1.11.5

package time

import original "time"
import "scrigo"
import "reflect"

var Package = scrigo.Package{
	"After": original.After,
	"AfterFunc": original.AfterFunc,
	"Date": original.Date,
	"Duration": reflect.TypeOf(original.Duration(int64(0))),
	"FixedZone": original.FixedZone,
	"LoadLocation": original.LoadLocation,
	"LoadLocationFromTZData": original.LoadLocationFromTZData,
	"Local": &original.Local,
	"Location": reflect.TypeOf(original.Location{}),
	"Month": reflect.TypeOf(original.Month(int(0))),
	"NewTicker": original.NewTicker,
	"NewTimer": original.NewTimer,
	"Now": original.Now,
	"Parse": original.Parse,
	"ParseDuration": original.ParseDuration,
	"ParseError": reflect.TypeOf(original.ParseError{}),
	"ParseInLocation": original.ParseInLocation,
	"Since": original.Since,
	"Sleep": original.Sleep,
	"Tick": original.Tick,
	"Ticker": reflect.TypeOf(original.Ticker{}),
	"Time": reflect.TypeOf(original.Time{}),
	"Timer": reflect.TypeOf(original.Timer{}),
	"UTC": &original.UTC,
	"Unix": original.Unix,
	"Until": original.Until,
	"Weekday": reflect.TypeOf(original.Weekday(int(0))),
}
