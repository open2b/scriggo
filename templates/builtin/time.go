// Copyright (c) 2021 Open2b Software Snc. All rights reserved.
// https://www.open2b.com

// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package builtin

import (
	"fmt"
	"time"

	"github.com/open2b/scriggo/templates"
)

// A Time represents an instant in time.
//
// It is a simplified version of the time.Time type of Go, suitable for use in
// a template.
type Time struct {
	t time.Time
	f TimeFormatter
}

// NewTime returns a Time value at the time of t, formatted using the
// formatter f. It is not intended to be used as a builtin but it can be used
// to implement builtins that return time values.
//
// To format time values as the time package of Go does, use nil as formatter
//
//    NewTime(t, nil)
//
// For example, a builtin that returns the current local time rounded to
// milliseconds and with location "America/New_York" can be implemented as
//
//    func now() builtin.Time {
//        l := time.LoadLocation("America/New_York")
//        t := time.Now().Rounded(time.Millisecond).In(l)
//        return builtin.NewTime(t, nil)
//    }
//
// and once added to the declarations
//
//    templates.Declarations{
//        ...
//        "now" : now,
//        ...
//    }
//
// can be used in a template
//
//    Current time: {{ now() }}
//
func NewTime(t time.Time, f TimeFormatter) Time {
	return Time{t, f}
}

// TimeArguments returns the time.Time and the formatter used as arguments for
// NewTime when t was created. It is not intended to be used as a builtin.
func TimeArguments(t Time) (time.Time, TimeFormatter) {
	return t.t, t.f
}

// TimeFormatter represents time formatters. It is not intended to be used as
// a builtin but for time formatter to pass to the NewTime function.
type TimeFormatter interface {
	// Format returns a textual representation of t formatted according to layout.
	// Format is called by the Time.Format method with a non-empty layout and by
	// the Time.String method with an empty layout.
	Format(t time.Time, layout string) string
}

// AddClock returns the time corresponding to adding the given number of
// hours, minutes and seconds to t. For example, AddClock(1, 30, 15) applied
// to January 1, 2011 14:10:20 returns January 1, 2011 15:40:35.
//
// AddClock normalizes its result in the same way that the Date function does.
func (t Time) AddClock(hours, minutes, seconds int) Time {
	d := time.Duration(hours)*time.Hour + time.Duration(minutes)*time.Minute + time.Duration(seconds)*time.Second
	return Time{t.t.Add(d), t.f}
}

// AddFraction returns the time corresponding to adding the given number of
// milliseconds, microseconds and nanoseconds to t.
//
// AddFraction normalizes its result in the same way that the Date function
// does.
func (t Time) AddFraction(msec, usec, nsec int) Time {
	d := time.Duration(msec)*time.Millisecond + time.Duration(usec)*time.Microsecond + time.Duration(nsec)*time.Nanosecond
	return Time{t.t.Add(d), t.f}
}

// AddDate returns the time corresponding to adding the given number of years,
// months and days to t. For example, AddDate(-1, 2, 3) applied to
// January 1, 2011 returns March 4, 2010.
//
// AddDate normalizes its result in the same way that the Date function does.
func (t Time) AddDate(years, months, days int) Time {
	return Time{t.t.AddDate(years, months, days), t.f}
}

// After reports whether the time instant t is after u.
func (t Time) After(u Time) bool {
	return t.t.After(u.t)
}

// Before reports whether the time instant t is before u.
func (t Time) Before(u Time) bool {
	return t.t.Before(u.t)
}

// Clock returns the hour, minute and second within the day specified by t.
// hour is in the range [0, 23], minute and second are in the range [0, 59].
func (t Time) Clock() (hour, minute, second int) {
	return t.t.Clock()
}

// Date returns the year, month and day in which t occurs. month is in the
// range [1, 12] and day is in the range [1, 31].
func (t Time) Date() (year, month, day int) {
	y, m, d := t.t.Date()
	return y, int(m), d
}

// Day returns the day of the month specified by t, in the range [1, 31]
func (t Time) Day() int {
	return t.t.Day()
}

// Equal reports whether t and u represent the same time instant.
func (t Time) Equal(u Time) bool {
	return t.t.Equal(u.t)
}

// Format formats the time value formatted according to layout using the t's
// formatter.
//
// As a special case, if layout is an empty string, it formats the time as for
// the time.RFC3339Nano layout of Go.
func (t Time) Format(layout string) string {
	if layout == "" {
		return t.t.Format(time.RFC3339Nano)
	}
	if t.f == nil {
		return t.t.Format(layout)
	}
	return t.f.Format(t.t, layout)
}

// Fraction returns the fractional seconds as milliseconds, microseconds and
// nanoseconds within the day specified by t.
//
// For example if the fractional seconds are 0.918273645, it returns 918,
// 918273 and 918273645.
func (t Time) Fraction() (msec, usec, nsec int) {
	d := time.Duration(t.t.Nanosecond())
	return int(d.Milliseconds()), int(d.Microseconds()), int(d.Nanoseconds())
}

// Hour returns the hour within the day specified by t, in the range [0, 23].
func (t Time) Hour() int {
	return t.t.Hour()
}

// JS returns the time as a JavaScript date. The result is undefined if the
// year of t is not in the range [-999999, 999999].
func (t Time) JS() templates.JS {
	y := t.t.Year()
	ms := int64(t.t.Nanosecond()) / int64(time.Millisecond)
	name, offset := t.t.Zone()
	if name == "UTC" {
		format := `new Date("%0.4d-%0.2d-%0.2dT%0.2d:%0.2d:%0.2d.%0.3dZ")`
		if y < 0 || y > 9999 {
			format = `new Date("%+0.6d-%0.2d-%0.2dT%0.2d:%0.2d:%0.2d.%0.3dZ")`
		}
		return templates.JS(fmt.Sprintf(format, y, t.t.Month(), t.t.Day(), t.t.Hour(), t.t.Minute(), t.t.Second(), ms))
	}
	zone := offset / 60
	h, m := zone/60, zone%60
	if m < 0 {
		m = -m
	}
	format := `new Date("%0.4d-%0.2d-%0.2dT%0.2d:%0.2d:%0.2d.%0.3d%+0.2d:%0.2d")`
	if y < 0 || y > 9999 {
		format = `new Date("%+0.6d-%0.2d-%0.2dT%0.2d:%0.2d:%0.2d.%0.3d%+0.2d:%0.2d")`
	}
	return templates.JS(fmt.Sprintf(format, y, t.t.Month(), t.t.Day(), t.t.Hour(), t.t.Minute(), t.t.Second(), ms, h, m))
}

// JSON returns the time as a JSON date.
func (t Time) JSON() templates.JSON {
	return templates.JSON(`"` + t.t.Format(time.RFC3339) + `"`)
}

// Month returns the month of the year specified by t, in the range [1, 12]
// where 1 is January and 12 is December.
func (t Time) Month() int {
	return int(t.t.Month())
}

// Minute returns the minute offset within the hour specified by t, in the range [0, 59].
func (t Time) Minute() int {
	return t.t.Minute()
}

// Second returns the second offset within the minute specified by t, in the range [0, 59].
func (t Time) Second() int {
	return t.t.Second()
}

// String returns the time formatted.
func (t Time) String() string {
	if t.f == nil {
		return t.t.Format(time.RFC3339Nano)
	}
	return t.f.Format(t.t, "")
}

// Sub returns t-u as the hours, minutes, seconds and nanoseconds. minutes and
// seconds are in the range [-59,59] and ns is in the range [0, 999999999].
//
// For example, Sub applied to March 27, 2021 11:21:14.964553 and March 27,
// 2021 10:16:29.370149 returns 1033h 4m 45s 594404000ns.
//
// t-u truncated, in seconds is
//    hours * 3600 + minutes * 60 + seconds
// t-u truncated, in minutes is
//    hours * 60 + minutes
// t-u truncated, in hours is
//    hours
//
func (t Time) Sub(u Time) (hours, minutes, seconds, nsec int) {
	d := t.t.Sub(u.t)
	h, d := d/time.Hour, d%time.Hour
	m, d := d/time.Minute, d%time.Minute
	s, d := d/time.Second, d%time.Second
	return int(h), int(m), int(s), int(d)
}

// UTC returns t with the location set to UTC.
func (t Time) UTC() Time {
	return Time{t.t.UTC(), t.f}
}

// Unix returns t as a Unix time, the number of seconds elapsed since
// January 1, 1970 UTC.
func (t Time) Unix() int64 {
	return t.t.Unix()
}

// Weekday returns the day of the week specified by t, in the range [0, 6]
// where 0 is Sunday and 6 is Saturday.
func (t Time) Weekday() int {
	return int(t.t.Weekday())
}

// Year returns the year in which t occurs.
func (t Time) Year() int {
	return t.t.Year()
}

// timeLayouts contains predefined layouts used by the ParseTime function.
//
// This list of layouts is the same used by Hugo's time function. It is
// copyright 2017 The Hugo Authors and licensed under the Apache license.
// See https://github.com/gohugoio/hugo/blob/master/tpl/time/time.go.
var timeLayouts = []string{
	time.RFC3339,
	"2006-01-02T15:04:05", // iso8601 without timezone
	time.RFC1123Z,
	time.RFC1123,
	time.RFC822Z,
	time.RFC822,
	time.RFC850,
	time.ANSIC,
	time.UnixDate,
	time.RubyDate,
	"2006-01-02 15:04:05.999999999 -0700 MST", // Time.String()
	"2006-01-02",
	"02 Jan 2006",
	"2006-01-02T15:04:05-0700", // RFC3339 without timezone hh:mm colon
	"2006-01-02 15:04:05 -07:00",
	"2006-01-02 15:04:05 -0700",
	"2006-01-02 15:04:05Z07:00", // RFC3339 without T
	"2006-01-02 15:04:05Z0700",  // RFC3339 without T or timezone hh:mm colon
	"2006-01-02 15:04:05",
	time.Kitchen,
	time.Stamp,
	time.StampMilli,
	time.StampMicro,
	time.StampNano,
}
