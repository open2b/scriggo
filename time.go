// Copyright (c) 2019 Open2b Software Snc. All rights reserved.
// https://www.open2b.com

// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package scrigo

import (
	"reflect"
	"time"
)

var _time = Package{
	"ANSIC":                  time.ANSIC,
	"UnixDate":               time.UnixDate,
	"RubyDate":               time.RubyDate,
	"RFC822":                 time.RFC822,
	"RFC822Z":                time.RFC822Z,
	"RFC850":                 time.RFC850,
	"RFC1123":                time.RFC1123,
	"RFC1123Z":               time.RFC1123Z,
	"RFC3339":                time.RFC3339,
	"RFC3339Nano":            time.RFC3339Nano,
	"Kitchen":                time.Kitchen,
	"Stamp":                  time.Stamp,
	"StampMilli":             time.StampMilli,
	"StampMicro":             time.StampMicro,
	"StampNano":              time.StampNano,
	"Nanosecond":             time.Nanosecond,
	"Microsecond":            time.Microsecond,
	"Millisecond":            time.Millisecond,
	"Second":                 time.Second,
	"Minute":                 time.Minute,
	"Hour":                   time.Hour,
	"January":                time.January,
	"February":               time.February,
	"March":                  time.March,
	"April":                  time.April,
	"May":                    time.May,
	"June":                   time.June,
	"July":                   time.July,
	"August":                 time.August,
	"September":              time.September,
	"October":                time.October,
	"November":               time.November,
	"December":               time.December,
	"Sunday":                 time.Sunday,
	"Monday":                 time.Monday,
	"Tuesday":                time.Tuesday,
	"Wednesday":              time.Wednesday,
	"Thursday":               time.Thursday,
	"Friday":                 time.Friday,
	"Saturday":               time.Saturday,
	"After":                  time.After,
	"Sleep":                  time.Sleep,
	"Tick":                   time.Tick,
	"ParseDuration":          time.ParseDuration,
	"Since":                  time.Since,
	"Until":                  time.Until,
	"FixedZone":              time.FixedZone,
	"LoadLocation":           time.LoadLocation,
	"LoadLocationFromTZData": time.LoadLocationFromTZData,
	"NewTicker":              time.NewTicker,
	"Date":                   time.Date,
	"Now":                    time.Now,
	"Parse":                  time.Parse,
	"ParseInLocation":        time.ParseInLocation,
	"Unix":                   time.Unix,
	"AfterFunc":              time.AfterFunc,
	"NewTimer":               time.NewTimer,
	"Duration":               reflect.TypeOf(time.Duration(0)),
	"Location":               reflect.TypeOf(time.Location{}),
	"Month":                  reflect.TypeOf(time.Month(0)),
	"Weekday":                reflect.TypeOf(time.Weekday(0)),
	"Timer":                  reflect.TypeOf(time.Timer{}),
	"Ticker":                 reflect.TypeOf(time.Ticker{}),
	"Time":                   reflect.TypeOf(time.Time{}),
	"ParseError":             reflect.TypeOf(time.ParseError{}),
}
