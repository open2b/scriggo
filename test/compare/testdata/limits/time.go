//+build !windows

// paniccheck -time=10ms

package main

import "time"

func main() {
	time.Sleep(3 * time.Millisecond)
}
