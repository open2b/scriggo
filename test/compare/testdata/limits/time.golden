panic: runtime error: out of time: context deadline exceeded